elasticsearch/src/main/java/org/hibernate/search/elasticsearch/util/impl/GsonHttpEntity.java

```diff
@@ -0,0 +1,205 @@
+package org.hibernate.search.elasticsearch.util.impl;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.Writer;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.util.List;
+import java.util.Objects;
+
+import org.apache.http.Header;
+import org.apache.http.HttpEntity;
+import org.apache.http.entity.ContentType;
+import org.apache.http.message.BasicHeader;
+import org.apache.http.nio.ContentEncoder;
+import org.apache.http.nio.IOControl;
+import org.apache.http.nio.entity.HttpAsyncContentProducer;
+import org.apache.http.protocol.HTTP;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonIOException;
+import com.google.gson.JsonObject;
+import com.google.gson.internal.Streams;
+import com.google.gson.stream.JsonWriter;
+
+/**
+ * Optimised adapter to encode GSON objects into HttpEntity instances.
+ * The naive approach was using various StringBuilders and the kind
+ * of objects we need to serialize into JSON where causing the internal
+ * StringBuilder buffers to need frequent resizing.
+ *
+ * Rather than try guessing reasonable default sizes for these buffers,
+ * we can defer the serialization to write directly into the ByteBuffer
+ * of the HTTP client. By doing so we can use buffer pages of fixed size.
+ *
+ * To implement this trick we have to implement both interfaces.
+ *
+ * @author Sanne Grinovero
+ */
+final class GsonHttpEntity implements HttpEntity, HttpAsyncContentProducer {
+
+	/**
+	 * A conservative quess at the size of the buffer in characters
+	 * needed to hold the produced JSON document.
+	 * Picking a too low size will cause CPU overhead, a too large size
+	 * will allocate unnecessarily large buffers (waste memory).
+	 * The heap consumption is short lived.
+	 */
+	private static final int WRITE_BUFFER_SIZE = 512;
+
+	private static final BasicHeader CONTENT_TYPE = new BasicHeader( HTTP.CONTENT_TYPE, ContentType.APPLICATION_JSON.toString() );
+
+	private final Gson gson;
+	private final List<JsonObject> bodyParts;
+
+	GsonHttpEntity(Gson gson, List<JsonObject> bodyParts) {
+		Objects.requireNonNull( gson );
+		Objects.requireNonNull( bodyParts );
+		this.gson = gson;
+		this.bodyParts = bodyParts;
+	}
+
+	@Override
+	public boolean isRepeatable() {
+		return true;
+	}
+
+	@Override
+	public boolean isChunked() {
+		return false;
+	}
+
+	@Override
+	public long getContentLength() {
+		//Can't computer the size in advance w/o it being very expensive
+		return -1;
+	}
+
+	@Override
+	public Header getContentType() {
+		return CONTENT_TYPE;
+	}
+
+	@Override
+	public Header getContentEncoding() {
+		return null;
+	}
+
+	@Override
+	public InputStream getContent() {
+		//This could be implemented but would be inefficient.
+		//I therefore prefer throwing the exception so that we can easily spot unintended usage via tests.
+		throw new UnsupportedOperationException( "Not implemented! Expected to produce content only over produceContent()" );
```

Comment:
Did you test this with elasticsearch-aws? I think we use this method to get the request body and hash it... Obviously we may want to also optimize AWS request hashing (maybe add a wrapper to the content encoder to hash on the fly), but my point is this may break AWS support.